File: ruby/kindergarten-garden/kindergarten_garden.rb
class Garden
  STUDENTS = %w(Alice Bob Charlie David Eve Fred Ginny Harriet Ileana Joseph Kincaid Larry)
  PLANTS = {"C" => :clover, "G" => :grass, "R" => :radishes, "V" => :violets}

  def initialize(diagram, students = STUDENTS)
    @diagram = diagram
    @students = students.map(&:downcase).sort.map(&:to_sym)
  end
  def method_missing(m)
    @diagram
      .split
      .flat_map { _1.chars.each_slice(2).to_a[@students.index(m)] }
      .map { PLANTS[_1] }
  end
end

File: clojure/kindergarten-garden/src/kindergarten_garden.clj
(ns kindergarten-garden
  (:require [clojure.string :as str]))

(def default-student-list
  ["Alice" "Bob" "Charlie" "David"
   "Eve" "Fred" "Ginny" "Harriet"
   "Ileana" "Joseph" "Kincaid" "Larry"])

(def plant-assoc {\R :radishes \C :clover \V :violets \G :grass})

(defn garden
  ([s] (garden s default-student-list))
  ([s students]
   (let [student-keys (->> students
              (map str/lower-case)
              (sort)
              (map keyword))
         garden-keys (->> s
              (str/split-lines)
              (map #(partition 2 %))
              (apply interleave)
              (flatten)
              (map plant-assoc)
              (partition 4))]
     (zipmap student-keys garden-keys))))

